exonum/src/blockchain/transaction.rs

```diff
@@ -0,0 +1,398 @@
+/// ```
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub enum TransactionError {
+    /// Panic occurred during transaction execution.
+    Panic,
+    /// General failure (unspecified reason).
+    Failure,
+    /// User-defined error-code. Can have different meanings for different transactions and
+    /// services.
+    Code(u8),
+    /// User-defined string error description.
+    Description(String),
+}
+
+/// Converts from `ExecutionResult` into `TransactionResult`.
+pub fn convert_result(result: ExecutionResult) -> TransactionResult {
+    match result {
+        Ok(()) => Ok(()),
+        Err(ExecutionError::Failure) => Err(TransactionError::Failure),
+        Err(ExecutionError::Code(c)) => Err(TransactionError::Code(c)),
+        Err(ExecutionError::Description(s)) => Err(TransactionError::Description(s)),
+    }
+}
+
+impl<'a, T: Transaction> From<T> for Box<Transaction + 'a> {
+    fn from(tx: T) -> Self {
+        Box::new(tx) as Box<Transaction>
+    }
+}
+
+impl fmt::Display for TransactionError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            TransactionError::Panic => write!(f, "Panic during execution"),
+            TransactionError::Failure => write!(f, "Unspecified failure"),
+            TransactionError::Code(c) => write!(f, "Error code: {}", c),
+            TransactionError::Description(ref s) => write!(f, "{}", s),
+        }
+    }
+}
+
+// String content (`TransactionError::Description`) is intentionally excluded from the hash
+// calculation because user can be tempted to use error description from a third-party libraries
+// which aren't stable across the versions.
+impl CryptoHash for TransactionResult {
+    fn hash(&self) -> Hash {
+        u16::hash(&status_as_u16(self))
+    }
+}
+
+// `TransactionResult` is stored as `u16` with optional string part needed only for
+// `TransactionError::Description`.
+impl StorageValue for TransactionResult {
+    fn into_bytes(self) -> Vec<u8> {
+        let mut res = u16::into_bytes(status_as_u16(&self));
+        if let Err(TransactionError::Description(s)) = self {
+            res.extend(String::into_bytes(s));
+        }
+        res
+    }
+
+    fn from_bytes(bytes: Cow<[u8]>) -> Self {
+        let main_part = u16::from_bytes(Cow::Borrowed(&bytes));
+        match main_part {
+            value @ 0...MAX_ERROR_CODE => Err(TransactionError::Code(value as u8)),
+            TRANSACTION_STATUS_OK => Ok(()),
+            TRANSACTION_STATUS_PANIC => Err(TransactionError::Panic),
+            TRANSACTION_STATUS_FAILURE => Err(TransactionError::Failure),
+            TRANSACTION_STATUS_DESCRIPTION => {
+                Err(TransactionError::Description(
+                    String::from_bytes(Cow::Borrowed(&bytes[2..])),
+                ))
+            }
+            value => panic!("Invalid TransactionResult value: {}", value),
+        }
+    }
+}
+
+fn status_as_u16(status: &TransactionResult) -> u16 {
```

Comment:
Why isn't this a private method in `TransactionResult`?

Reply:
I don't know. :upside_down_face: In C++ it is often considered that free functions should be preferred if possible. I'm not sure about the best practices for Rust in this case.

@aleksuss @matklad @vitvakatu @ozkriff Opinions?

Reply:
- I am not sure that there are best practices for Rust for this particular case
- In C++, the rational behind the rule is that free functions do not have access to private members, and this is not applicable to Rust, which uses modules for visibility. 
- I personally generally prefer free-standing functions to methods and traits most of the times, because it the simplest tool of abstraction. That is, I prefer C-style Rust to C++-style Rust for implementation parts of code.
- However, in this particular case I would make it a method: this will allow to get rid of the `status` word from the function name :)

Reply:
I don't have strong opinion, but I'd prefer to see this function as method.

Reply:
I've used [this](https://doc.rust-lang.org/1.0.0/style/features/functions-and-methods/README.html) as a reference.

Reply:
OK, now I'm convinced, but Rust complains again:
> cannot define inherent `impl` for a type outside of the crate where the type is defined

I don't think that creating a method for `TransactionError` and reusing it in a (still needed) free function makes code better.

Reply:
Well, it could be possible to implement a private trait to convert `TransactionResult` as needed, something like this:

```rust
trait AsCode {
    type Code: StorageValue + CryptoHash;
    fn as_code(&self) -> Self::Code;
}

impl AsCode for TransactionResult {
    type Code = u16;

    fn as_code(&self) -> u16 {
        match *self {
            Ok(()) => TRANSACTION_STATUS_OK,
            Err(TransactionError::Panic) => TRANSACTION_STATUS_PANIC,
            Err(TransactionError::Failure) => TRANSACTION_STATUS_FAILURE,
            Err(TransactionError::Code(c)) => u16::from(c),
            Err(TransactionError::Description(_)) => TRANSACTION_STATUS_DESCRIPTION,
        }
    }
}

// usage
impl CryptoHash for TransactionResult {
    fn hash(&self) -> Hash {
        self.as_code().hash()
    }
}
```

Not sure whether it would be any better than the current implementation, though.